#!/bin/sh -e

fatal() { echo "FATAL: $@" 1>&2; exit 1; }

lucid_patch() {
patch -f /usr/bin/dpkg-gencontrol << 'EOF'
--- /usr/bin/dpkg-gencontrol.orig       2010-04-15 17:23:43.000000000 +0000
+++ /usr/bin/dpkg-gencontrol    2010-06-03 12:11:23.000000000 +0000
@@ -291,6 +291,7 @@
     if (!$c) {
         chdir("$packagebuilddir") ||
             syserr(_g("chdir for du to \`%s'"), $packagebuilddir);
+        system("find -type f | xargs touch"); # workaround for evil aufs bug
         exec("du", "-k", "-s", ".") or syserr(_g("exec %s"), "du");
     }
     my $duo = '';
EOF
}

lenny_patch() {
patch -f /usr/bin/dpkg-gencontrol << 'EOF'
--- /usr/bin/dpkg-gencontrol.orig       2010-06-23 17:51:35.000000000 +0000
+++ /usr/bin/dpkg-gencontrol    2010-06-23 17:52:28.000000000 +0000
@@ -288,6 +288,7 @@
     if (!$c) {
         chdir("$packagebuilddir") ||
             syserr(_g("chdir for du to \`%s'"), $packagebuilddir);
+        system("find -type f | xargs touch"); # workaround for evil aufs bug
         exec("du","-k","-s",".") or &syserr(_g("exec du"));
     }
     my $duo = '';
EOF
}

squeeze_patch() {
patch -f /usr/bin/dpkg-gencontrol << 'EOF'
--- dpkg-gencontrol.orig        2012-02-16 11:15:44.000000000 +0000
+++ dpkg-gencontrol     2012-02-16 11:16:18.000000000 +0000
@@ -301,6 +301,7 @@
     if (!$c) {
         chdir("$packagebuilddir") ||
             syserr(_g("chdir for du to \`%s'"), $packagebuilddir);
+        system("find -type f | xargs touch"); # workaround for evil aufs bug
         exec("du", "-k", "-s", ".") or syserr(_g("exec %s"), "du");
     }
     my $duo = '';
EOF
}

case "$CODENAME" in
    lucid) lucid_patch ;;
    lenny) lenny_patch ;;
    squeeze) squeeze_patch ;;
    *) fatal "CODENAME is not set or not recognized" ;;
esac

